metricbeat/mb/parse/url.go

```diff
@@ -189,8 +194,62 @@ func getURL(rawURL, scheme, username, password, path, query string) (*url.URL, e
 	}
 
 	SetURLUser(u, username, password)
+	// url == unix:///tmp/foo.sock
+
+	if strings.HasSuffix(u.Scheme, "unix") || strings.HasSuffix(u.Scheme, "npipe") {
+		// Taking the URI RFC (RFC3986) https://tools.ietf.org/html/rfc3986 as the source of truth.
+		//
+		// The specs does not contains an official way of defining an absolute HTTP resource
+		// while using a unix domain socket or a npipe as the transport layer.
+		// Lets say that you want to connect to the socket and get a path. (GET /stats)
+		// in the a normal URI you would define it like this http://host/stats so the path and the host
+		// would be parsed in two different field like this:
+		//
+		// {
+		// Scheme: "http",
+		// Host: "host",
+		// Path: "stats",
+		// }
+		//
+		// Now if you take the Golang URL parser and use an URI to define a unix domain socket:
+		// unix:///tmp/fb.sock, the parser will do the following.
+		//
+		// {
+		// Scheme: "unix",
+		// Host: "",
+		// Path: "/tmp/fb.sock"
+		// }
+		//
+		// Now lets say that you want to access a specific path after making the connection? You cannot
+		// set that information in the path. The other choice would be to take the path and use it as the
+		// host and set the path to the actual resource. The problem is if you do that, the path will be
+		// URL encoded. Now the information is correctly set in the url.URL struct. But if you take the
+		// object and convert it back to a string and later try to parse it again. The URI will now
+		// be invalid because of the encoding. This will also breaks multiples modules tests. So I think
+		// The easiest way to go around that is to define an internal DSN when a unix domain socket is
+		// used and set the path as a query string. Taking the socket file defined earlier and the "stats"
+		// path the struct will look like this.
+		//
+		// {
+		// Scheme: "unix",
+		// Host: "",
+		// Path: "/tmp/fb.sock?__path=stats
+		// }
+		//
+		// The change will not impact any existing behavior that use a parsed sockfile directly, but if
+		// you use the socket file to connect via HTTP it will be transparent.
+		if u.Host == "" && path != "" {
```

Comment:
I don't like this.. but I  haven't found a better way without changing a lot of stuff/behavior. We are also a bit on our own with the specs concerning URI definition. And This is really linked to our API definition in the metricbeat HTTP client that actually get accepts the full URI instead of configuring the initial client with the transport (or host) then executing query. Looking around the internet for what others did, I saw a single use case for postgresql that were splitting the path the socket using the query string. And using an encoded hosts which we cannot do.

Reply:
can you link postres solution?

Reply:
Can't find the actual link but the solution was looking like this:

`postgresql:///tmp/sock?path=X`

Reply:
closing this, I've changed strategy.